apex/contrib/group_norm/group_norm.py

```diff
@@ -0,0 +1,258 @@
+                G,
+                weight,
+                bias,
+                eps,
+                act="",
+                algo=group_norm_cuda.OnePass):
+        # sanity check
+        act = act.lower()
+        assert x.is_contiguous(memory_format=torch.channels_last), \
+            "Only support NHWC layout."
+        assert weight.numel() == x.shape[1], "Unexpected parameter count."
+        assert bias.numel() == x.shape[1], "Unexpected parameter count."
+        assert x.shape[1] % G == 0, "C % G != 0."
+        assert act in ["", "silu", "swish"], "Unsupported activation."
+
+        with_swish = (act in ["silu", "swish"])
+
+        # enqueue fprop kernel
+        y, sums = group_norm_cuda.forward(x, G, weight, bias, eps, algo,
+                                          with_swish)
+
+        # save for backward
+        ctx.save_for_backward(x, weight, bias, sums)
+        ctx.G = G
+        ctx.eps = eps
+        ctx.algo = algo
+        ctx.with_swish = with_swish
+        return y
+
+    @staticmethod
+    def backward(ctx, dy):
+        # sanity check
+        assert dy.is_contiguous(memory_format=torch.channels_last), \
+            "Only support NHWC layout."
+
+        # retrive saved info
+        x, w, b, sums = ctx.saved_tensors
+        G = ctx.G
+        eps = ctx.eps
+        algo = ctx.algo
+        with_swish = ctx.with_swish
+
+        # enqueue bprop kernel
+        dx, dw, db = group_norm_cuda.backward(dy, sums, x, G, w, b, eps, algo,
+                                              with_swish)
+
+        return dx, None, dw, db, None, None, None
+
+
+class GroupNormOnePass(GroupNormNHWC):
+
+    @staticmethod
+    @disable
+    def forward(ctx, x, G, weight, bias, eps, act=""):
+        return super(GroupNormOnePass,
+                     GroupNormOnePass).forward(ctx, x, G, weight, bias, eps,
+                                               act, group_norm_cuda.OnePass)
+
+
+class GroupNormTwoPass(GroupNormNHWC):
+
+    @staticmethod
+    @disable
+    def forward(ctx, x, G, weight, bias, eps, act=""):
+        return super(GroupNormTwoPass,
+                     GroupNormTwoPass).forward(ctx, x, G, weight, bias, eps,
+                                               act, group_norm_cuda.TwoPass)
+
+
+cuda_group_norm_nhwc = GroupNormNHWC.apply
+cuda_group_norm_nhwc_one_pass = GroupNormOnePass.apply
+cuda_group_norm_nhwc_two_pass = GroupNormTwoPass.apply
+
+
+# We do not direct inherit from torch.nn.GroupNorm since several fusers don't
+# support inheritance. Extends:
+# https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/normalization.py
+class GroupNorm(torch.nn.Module):
+    """Optimized GroupNorm for NHWC layout with optional Swish/SiLU fusion.
+
+    There are two version of CUDA kernels under the hood: one pass and two
+    passes. This operator contains a simple heuristic to choose algorithm.
+
+    Limitations:
+
+    * Designed for 32 groups, though some other number of groups can also work;
```

Comment:
Shouldn't 16 be mentioned here as well?

Reply:
if this is intended for 16/32, then I'd mention those values and also mention that this will call pytorch based implementation for the unsupported groups values

Reply:
You're right, that makes sense and I just updated the doc for the module.